src/can/mod.rs

```diff
@@ -45,3 +45,71 @@ pub trait Frame: Sized {
     /// Returns the frame data (0..8 bytes in length).
     fn data(&self) -> &[u8];
 }
+
+/// CAN error
+pub trait Error: core::fmt::Debug {
+    /// Convert error to a generic CAN error kind
+    ///
+    /// By using this method, CAN errors freely defined by HAL implementations
+    /// can be converted to a set of generic serial errors upon which generic
+    /// code can act.
+    fn kind(&self) -> ErrorKind;
+}
+
+/// CAN error kind
+///
+/// This represents a common set of CAN operation errors. HAL implementations are
+/// free to define more specific or additional error types. However, by providing
+/// a mapping to these common CAN errors, generic code can still react to them.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+#[non_exhaustive]
+pub enum ErrorKind {
+    /// The peripheral receive buffer was overrun.
+    Overrun,
+
+    // MAC sublayer errors
+    /// A bit error is detected at that bit time when the bit value that is
+    /// monitored differs from the bit value sent.
+    Bit,
+
+    /// A stuff error is detected at the bit time of the sixth consecutive
+    /// equal bit level in a frame field that shall be coded by the method
+    /// of bit stuffing.
+    Stuff,
+
+    /// Calculated CRC sequence does not equal the received one.
+    Crc,
+
+    /// A form error shall be detected when a fixed-form bit field contains
+    /// one or more illegal bits.
+    Form,
+
+    /// An ACK  error shall be detected by a transmitter whenever it does not
+    /// monitor a dominant bit during the ACK slot.
+    Ack,
```

Comment:
For no-acknowledge errors we selected `NoAcknowledge` over at #296 . It would be good to have the same name in both sides.

Reply:
The CAN 2.0 standard says there are 5 error types, which it calls:

* `BIT ERROR`
* `STUFF ERROR`
* `CRC ERROR`
* `FORM ERROR`
* `ACKNOWLEDGEMENT ERROR`

So matching those names seems nice, while in I²C the condition where there's no acknowledgement is explicitly called `NAK`/no acknowledge, so it makes sense to use NoAcknowledge for I²C. I guess there's a question of whether it should be `Ack` or `Acknowledgement` here, but given how long the latter is, I'd be fine with `Ack`...